src/dawn/CodeGen/Cuda/MSCodeGen.cpp

```diff
@@ -717,6 +719,40 @@ void MSCodeGen::generateCudaKernelCode() {
   if(containsTemporary && useTmpIndex_)
     fnDecl = "template<typename TmpStorage>";
   fnDecl = fnDecl + "__global__ void";
+
+  int maxThreadsPerBlock =
+      blockSize_[0] * (blockSize_[1] + maxExtents[1].Plus - maxExtents[1].Minus +
+                       (maxExtents[0].Minus < 0 ? 1 : 0) + (maxExtents[0].Plus > 0 ? 1 : 0));
+
+  std::string arch = context_->getOptions().arch;
+  std::string domain_size = context_->getOptions().domain_size;
+  // if we have information about the architecture and the domain sizes we can further specify the
+  // minimum number of blocks per SM of the GPU
+  if((arch == "P100" || arch == "K40" || arch == "K80") && !domain_size.empty()) {
+    std::istringstream idomain_size(domain_size);
+    std::string arg;
+    getline(idomain_size, arg, ',');
+    int isize = std::stoi(arg);
+    getline(idomain_size, arg, ',');
+    int jsize = std::stoi(arg);
+
+    int nSM = -1;
+    if(context_->getOptions().arch == "P100") {
+      nSM = 56;
+    } else if(arch == "K40" || arch == "K80") {
+      nSM = 15;
+    }
+
+    int minBlocksPerSM = isize * jsize / (blockSize_[0] * blockSize_[1]);
+    if(solveKLoopInParallel_)
+      minBlocksPerSM *= 80 / blockSize_[2];
+    minBlocksPerSM /= nSM;
+
+    fnDecl = fnDecl + " __launch_bounds__(" + std::to_string(maxThreadsPerBlock) + "," +
+             std::to_string(minBlocksPerSM) + ") ";
+  } else {
+    fnDecl = fnDecl + " __launch_bounds__(" + std::to_string(maxThreadsPerBlock) + ") ";
```

Comment:
if we keep it we want to exit with an illegal architecture and do this if nothing is specified

Reply:
the idea here was not to exit, but rather to not specify the min number of blocks per sm if we dont know how many sms we have. That should still be allowed

Reply:
Alright. I would advise to at least print it into the log though. If we have a typo in a script and are not getting the optimal block-size silently it might be a hassle

Reply:
I am not sure where to put this, we dont have a well structured logging (only the INFO is working). And I am not sure you want always to print a message that for most people here wont mean anything. 
In many cases you will compile an executable for which the domain sizes are not known, so it is not possible to fully specify the launch bounds, and that should be ok, performance would not be bad